### ISSUE TITLE
Cannot test JSONP HTTP request using the HttpClientTestingModule and HttpTestingController

### ISSUE THREAD

**Issue by courtneypattison** [angular/angular#20878]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

When I try to mock a JSONP HTTP request using the HttpClientTestingModule, I get the error

`Error: Expected one matching request for criteria "Match URL: https://archive.org/index.php?output=json&callback=callback", found none.`

I posted a [stackoverflow question](https://stackoverflow.com/questions/47703877/how-do-i-test-a-jsonp-http-request-in-angular) about this issue.

## Expected behavior

A TestRequest to be returned from `httpMock.expectOne(service.url);` in the following code sample.

## Minimal reproduction of the problem with instructions

I'm using the [HttpClientModule][1] and [HttpClientJsonpModule][2] to make a JSONP HTTP request in a service.

**app.module.ts**

    import { BrowserModule } from '@angular/platform-browser';
    import { NgModule } from '@angular/core';
    import { HttpClientModule, HttpClientJsonpModule } from '@angular/common/http';
    
    import { AppComponent } from './app.component';
    
    @NgModule({
      declarations: [
        AppComponent
      ],
      imports: [
        BrowserModule,
        HttpClientModule,
        HttpClientJsonpModule
      ],
      providers: [],
      bootstrap: [AppComponent]
    })
    export class AppModule { }

This service uses the [jsonp][3] method from the [HttpClient class][4] to get a JSONP response from the specified URL. 

**example.service.ts**

    import { Injectable } from '@angular/core';
    import { HttpClient } from '@angular/common/http';
    
    @Injectable()
    export class ExampleService {
      url = "https://archive.org/index.php?output=json&callback=callback";
    
      constructor(private http: HttpClient) { }
    
      getData() {
        return this.http.jsonp(this.url, 'callback');
      }
    }

Using the [HttpClientTestingModule][7], I inject the [HttpTestingController][8] so I can mock and flush my JSONP HTTP request.

**example.service.spec.ts**

    import { TestBed, inject } from '@angular/core/testing';
    import { 
      HttpClientTestingModule,
      HttpTestingController
    } from '@angular/common/http/testing';
    
    import { ExampleService } from './example.service';
    
    describe('ExampleService', () => {
      let service: ExampleService;
      let httpMock: HttpTestingController;
    
      beforeEach(() => {
        TestBed.configureTestingModule({
          imports: [HttpClientTestingModule],
          providers: [ExampleService]
        });
    
        service = TestBed.get(ExampleService);
        httpMock = TestBed.get(HttpTestingController);
      });
    
      describe('#getData', () => {
        it('should return an Observable<any>', () => {
          const dummyData = { id: 1 };
    
          service.getData().subscribe(data => {
            expect(data).toEqual(dummyData);
          });
    
          const req = httpMock.expectOne(service.url); // Error
          expect(req.request.method).toBe('JSONP');
          req.flush(dummyData);
        });
      });
    });

If I change the request method to GET, this test works as expected.

Also, on Gitter @alxhub showed me that I could get the expected return value from `httpMock.expectOne(service.url);` if I change `example.service.spec.ts` to the following:

    import { TestBed, inject } from '@angular/core/testing';
    import {
      HttpClientJsonpModule,
      HttpBackend,
      JsonpClientBackend
    } from '@angular/common/http';
    import { 
      HttpClientTestingModule,
      HttpTestingController
    } from '@angular/common/http/testing';
    
    import { ExampleService } from './example.service';
    
    describe('ExampleService', () => {
      let service: ExampleService;
      let httpMock: HttpTestingController;
    
      beforeEach(() => {
        TestBed.configureTestingModule({
          imports: [HttpClientTestingModule],
          providers: [ExampleService]
        });
    
        service = TestBed.get(ExampleService);
        httpMock = TestBed.get(HttpTestingController);
      });
    
      describe('#getData', () => {
        it('should return an Observable<any>', () => {
          const dummyData = { id: 1 };
    
          service.getData().subscribe(data => {
            expect(data).toEqual(dummyData);
          });
    
          const req = httpMock.expectOne(request => request.url === service.url);
          expect(req.request.method).toBe('JSONP');
          req.flush(dummyData);
        });
      });
    });

  [1]: https://angular.io/api/common/http/HttpClientModule
  [2]: https://angular.io/api/common/http/HttpClientJsonpModule
  [3]: https://angular.io/api/common/http/HttpClient#jsonp
  [4]: https://angular.io/api/common/http/HttpClient
  [5]: https://angular.io/api/common/http/JsonpInterceptor
  [6]: https://angular.io/api/common/http/JsonpClientBackend
  [7]: https://angular.io/api/common/http/testing/HttpClientTestingModule
  [8]: https://angular.io/api/common/http/testing/HttpTestingController

## What is the motivation / use case for changing the behavior?

To be able to test JSONP HTTP requests using the HttpClientTestingModule and HttpTestingController.

## Environment

<pre><code>
Angular version: 5.0.5

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.1  
- Platform:  Mac OS 10.13.1 

Others:

</code></pre>

**Comment by alxhub** [comment]
Yep, you're absolutely right - the mock doesn't really provide a good mechanism for expecting `JSONP` requests. It shouldn't be too hard to design something to handle that.
